src/math.rs

```diff
@@ -0,0 +1,182 @@
+    }
+
+    /// Construct with internal value derived from the system time
+    pub fn try_from_system_time(value: SystemTime, scale: f64) -> Result<Self, SystemTimeError> {
+        value
+            .elapsed()
+            .map(|t| BaseValue::new(t.as_secs_f64(), scale))
+    }
+
+    /// Get the inner value
+    pub fn value(&self) -> f64 {
+        self.value
+    }
+
+    /// Update the inner value
+    pub fn update(&mut self) {
+        self.value += PI * 2.0 * random_interval() * self.scale;
+    }
+}
+
+/// Math functions
+#[derive(Debug, Clone, Default, clap::ValueEnum)]
+pub enum MathFunction {
+    /// Sine
+    Sin,
+    /// Noise sine
+    NoiseSin,
+    /// Random spike sine
+    RandomSpikesSin,
+    /// Constant value
+    Const,
+    /// Saw
+    Saw,
+    /// Rect
+    Rect,
+    /// Sinc
+    Sinc,
+    /// Random value
+    Random,
+    /// Default value
+    #[default]
+    Default,
+}
+
+impl Display for MathFunction {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let s = match self {
+            MathFunction::Sin => "sin",
+            MathFunction::NoiseSin => "noise sin",
+            MathFunction::RandomSpikesSin => "random spikes sin",
+            MathFunction::Const => "const",
+            MathFunction::Saw => "saw",
+            MathFunction::Rect => "rect",
+            MathFunction::Sinc => "sinc",
+            MathFunction::Random => "random",
+            MathFunction::Default => "default",
+        };
+
+        write!(f, "{s}")
+    }
+}
+
+impl From<String> for MathFunction {
+    fn from(value: String) -> Self {
+        match value.as_str() {
+            "sin" => Self::Sin,
+            "noise-sin" => Self::NoiseSin,
+            "random-spikes-sin" => Self::RandomSpikesSin,
+            "const" => Self::Const,
+            "saw" => Self::Saw,
+            "rect" => Self::Rect,
+            "sinc" => Self::Sinc,
+            "random" => Self::Random,
+            _ => Self::Default,
+        }
+    }
+}
+
+impl MathFunction {
+    /// Use math functions to compute an output value given a certain input
+    pub fn compute(&self, value: f64) -> f64 {
+        match self {
+            MathFunction::Sin => sin(value),
+            MathFunction::NoiseSin => noise_sin(value),
+            MathFunction::RandomSpikesSin => random_spikes_sin(value),
+            MathFunction::Const => constant(value),
+            MathFunction::Saw => saw(value),
+            MathFunction::Rect => rect(value),
+            MathFunction::Sinc => sinc(value),
+            MathFunction::Random => random(),
+            MathFunction::Default => default(value),
+        }
+    }
+}
+
+fn sin(value: f64) -> f64 {
+    value.sin()
+}
+
+fn noise_sin(value: f64) -> f64 {
+    value.sin() + rand::random::<f64>() / f64::MAX * 0.2
+}
+
+fn random_spikes_sin(value: f64) -> f64 {
+    if value.sin() + rand::random::<f64>() / f64::MAX * 0.1 + rand::random::<f64>() / f64::MAX * 0.1
```

Comment:
Yep, the division by f64::MAX can be avoided

Reply:
Btw, if I understood correctly from the stream-qt5-test code, the _random_spike_sin_ function is a _spike_ (spike(x) = 1 if 0<=x<1, 0 otherwise) computed on the value of a _noise sine_ (which is sin(x) + rand())